### PR TITLE
System.Reflection.* missed Equals nullable annotations 

### PR DESCRIPTION
**Pull request by hrrrrustic**

### CODE REVIEW

**Comment by hrrrrustic** [comment]
Ugh, bot didn't ping anyone
@buyaa-n @joperezr @krwq @steveharter 
Found all of you in area owners for System.Reflection, sry if I ping someone incorrectly 
This PR is continuation of #47598

**Comment by hrrrrustic** [comment]
Btw there are some changes for reflection in #52167 under CoreLib and I'm not sure whom should I ping for that PR

File: src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Handle.cs [hunk 1]
```diff
@@ -159,7 +160,7 @@ internal int Token
             }
         }
 
-        public override bool Equals(object? obj)
+        public override bool Equals([NotNullWhen(true)] object? obj)
         {
             return obj is Handle && Equals((Handle)obj);
```
Comment (buyaa-n) [inline src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Handle.cs:165]: Please remove extra cast, please do same all below occurrences, not gonna comment more

**Review by buyaa-n — APPROVED** [review]
Thanks @hrrrrustic, could you remove extra castings in the implementation? Other than that LGTM.

**Comment by hrrrrustic** [comment]
Does it make sense to remove nullable marking here? https://github.com/dotnet/runtime/blob/3ba6ee15d7d212bedbf00e01d2125c44fb31e64b/src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/General/RoAssemblyName.cs#L52-L55

File: src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/PortablePdb/Handles.Debug.cs [hunk 2]
```diff
@@ -412,7 +413,7 @@ public bool IsNil
             return left._rowId == right._rowId;
         }
 
-        public override bool Equals(object? obj)
+        public override bool Equals([NotNullWhen(true)] object? obj)
         {
             return obj is LocalConstantHandle && ((LocalConstantHandle)obj)._rowId == _rowId;
```
Comment (buyaa-n) [inline src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/PortablePdb/Handles.Debug.cs:418]: One more left:
```suggestion
            return obj is LocalConstantHandle lch && lch._rowId == _rowId;
```

File: src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/PortablePdb/Handles.Debug.cs [hunk 3]
```diff
@@ -495,7 +496,7 @@ public bool IsNil
             return left._rowId == right._rowId;
         }
 
-        public override bool Equals(object? obj)
+        public override bool Equals([NotNullWhen(true)] object? obj)
         {
             return obj is ImportScopeHandle && ((ImportScopeHandle)obj)._rowId == _rowId;
```
Comment (buyaa-n) [inline src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/PortablePdb/Handles.Debug.cs:501]: Another here, please update

File: src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/PortablePdb/Handles.Debug.cs [hunk 4]
```diff
@@ -578,7 +579,7 @@ public bool IsNil
             return left._rowId == right._rowId;
         }
 
-        public override bool Equals(object? obj)
+        public override bool Equals([NotNullWhen(true)] object? obj)
         {
             return obj is CustomDebugInformationHandle && ((CustomDebugInformationHandle)obj)._rowId == _rowId;
```
Comment (buyaa-n) [inline src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/PortablePdb/Handles.Debug.cs:584]: same

File: src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeSystem/Handles.TypeSystem.cs [hunk 5]
```diff
@@ -2518,7 +2519,7 @@ internal StringHandle GetFullName()
             return StringHandle.FromOffset(GetHeapOffset());
         }
 
-        public override bool Equals(object? obj)
+        public override bool Equals([NotNullWhen(true)] object? obj)
         {
             return obj is NamespaceDefinitionHandle && Equals((NamespaceDefinitionHandle)obj);
```
Comment (buyaa-n) [inline src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeSystem/Handles.TypeSystem.cs:2524]: here

File: src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeSystem/Handles.TypeSystem.cs [hunk 6]
```diff
@@ -2649,7 +2650,7 @@ private ushort VirtualValue
             get { return unchecked((ushort)(_value >> 8)); }
         }
 
-        public override bool Equals(object? obj)
+        public override bool Equals([NotNullWhen(true)] object? obj)
         {
             return obj is BlobHandle && Equals((BlobHandle)obj);
```
Comment (buyaa-n) [inline src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeSystem/Handles.TypeSystem.cs:2655]: same

File: src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeSystem/Handles.TypeSystem.cs [hunk 7]
```diff
@@ -2717,7 +2718,7 @@ internal int Index
             get { return _index; }
         }
 
-        public override bool Equals(object? obj)
+        public override bool Equals([NotNullWhen(true)] object? obj)
         {
             return obj is GuidHandle && Equals((GuidHandle)obj);
```
Comment (buyaa-n) [inline src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeSystem/Handles.TypeSystem.cs:2723]: same

**Comment by buyaa-n** [comment]
> Does it make sense to remove nullable marking here?
> 
> https://github.com/dotnet/runtime/blob/3ba6ee15d7d212bedbf00e01d2125c44fb31e64b/src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/General/RoAssemblyName.cs#L52-L55

That makes sense, but I guess it has `?` because the implemented interface signature has `?`, if you remove that you will need to suppress the warning, by accounting that this is an internal type probably no need to remove it

**Review by buyaa-n — APPROVED** [review]
Looks good, thanks @hrrrrustic!

**Comment by jeffhandley** [comment]
@buyaa-n Once this is merged, please edit the description on https://github.com/dotnet/docs/issues/21202 to include the affected APIs from this PR.

**Comment by buyaa-n** [comment]
> @buyaa-n Once this is merged, please edit the description on dotnet/docs#21202 to include the affected APIs from this PR.

@jeffhandley I do not really think that is necessary, first, this is not breaking change and not really changing the annotation, just adding a `[NotNullWhen(true)]` for `Object.Equals(object obj)` overload which applies almost all types that overwrites `Object.Equals` method which is way too many to add into that issue (most of the annotations added [previously here](https://github.com/dotnet/runtime/pull/47598))

**Comment by jeffhandley** [comment]
I had noticed that the issue already mentioned #47598, so I thought we'd want to do something similar for other PRs. I don't think we need to call out each specific API. I wasn't sure if any of these could be considered breaking or not.

> dotnet/runtime#47598 [Breaking]
> [NotNullWhen(true)] added to hundreds of arguments

**Pull request merged**